### PR TITLE
Remove OAHashSet from merkle trees

### PR DESCRIPTION
**Pull request by mmedenjak**
EE: https://github.com/hazelcast/hazelcast-enterprise/pull/3537

`OAHashSet`s were collections in merkle tree leaves holding references to keys. It allows us to easily collect keys and entries when doing merkle tree sync. Unfortunately, these consume lots of heap, especially when the partition and map count is high. Instead, we will now iterate over the entire partition and calculate the hashcode to determine if the entry belongs to the leaf or not. Turns out, iterating the partition is not that costly.

Further heap cost optimizations of the tree are possible but are out of scope for this PR.

Results for iterating a single partition in a record store and calculating hash codes for a number of entries on a lab machine. Entry count is number of entries in a single partition. Key size is in bytes. Keys are always serialized so there's no point in measuring with `OBJECT` format or using classes with complicated `hashcode` implementations, the size of the byte array and the backing store implementation is all that matters. 

```
Benchmark                                               (entryCount)  (inMemoryFormat)  (keySize)  Mode  Cnt   Score   Error  Units
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            NATIVE         10  avgt   50   0.163 ? 0.003  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            NATIVE        128  avgt   50   0.163 ? 0.002  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            NATIVE       1024  avgt   50   0.163 ? 0.002  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            BINARY         10  avgt   50   1.376 ? 0.012  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            BINARY        128  avgt   50   1.366 ? 0.015  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration         10000            BINARY       1024  avgt   50   1.378 ? 0.014  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            NATIVE         10  avgt   50   1.543 ? 0.009  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            NATIVE        128  avgt   50   1.542 ? 0.015  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            NATIVE       1024  avgt   50   1.542 ? 0.008  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            BINARY         10  avgt   50  14.662 ? 0.227  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            BINARY        128  avgt   50  14.613 ? 0.201  ms/op
RecordStoreMerkleTreeIterationBenchmark.test_iteration        100000            BINARY       1024  avgt   50  14.359 ? 0.155  ms/op
```

We can see that the overhead of calculating hashcode is insignificant in relation to synchronizing partitions over WAN or local network, which means we can proceed with removing the `OAHashSet`.

### CODE REVIEW

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1230/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1230/Hazelcast-pr-builder-1230.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 7 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:22:8: Unused import - java.util.function.Consumer. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:95: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:96: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:97: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:98: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:99: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder_3/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:100: Don't use trailing comments. [TrailingComment]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1232/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1232/Hazelcast-pr-builder-1232.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There are 7 errors reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:22:8: Unused import - java.util.function.Consumer. [UnusedImports]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:95: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:96: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:97: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:98: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:99: Don't use trailing comments. [TrailingComment]
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:100: Don't use trailing comments. [TrailingComment]
--------------------------
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1243/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1243/Hazelcast-pr-builder-1243.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
-------TEST FAILURE-------
--------------------------
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   ArrayMerkleTreeTest.testFootprint:64
[ERROR]   ArrayMerkleTreeTest.testFootprintChanges:77
[INFO] 
[ERROR] Tests run: 27791, Failures: 2, Errors: 0, Skipped: 1001
[INFO] 

[ERROR] There are test failures.
</details>
</pre>

**Comment by hz-highfive** [comment]
The job `Hazelcast-pr-builder` of your PR failed ([log](
https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1280/build_log.txt), [artifacts](https://j-artifacts.s3.amazonaws.com/Hazelcast-pr-builder/1280/Hazelcast-pr-builder-1280.zip)). 
Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.
<details>
<summary>Click to expand the log file</summary>
<pre>
--------------------------
---------SUMMARY----------
--------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:3.1.1:checkstyle (default) on project hazelcast: An error has occurred in Checkstyle report generation. Failed during checkstyle execution: There is 1 error reported by Checkstyle 8.29 with /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/checkstyle/checkstyle.xml ruleset. -> [Help 1]
--------------------------
---------ERRORS-----------
--------------------------
[ERROR] /home/jenkins/jenkins_slave/workspace/Hazelcast-pr-builder/hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:22:8: Unused import - java.util.function.Consumer. [UnusedImports]
--------------------------
</details>
</pre>

File: hazelcast/src/test/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTreeTest.java [hunk 1]
```diff
@@ -67,11 +60,11 @@ public void testFootprint() {
             merkleTree2.updateAdd(i, i);
         }
 
-        assertTrue(merkleTree2.footprint() > merkleTree1.footprint());
+        assertEquals(merkleTree2.footprint(), merkleTree1.footprint());
```
Comment (mmedenjak) [inline hazelcast/src/test/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTreeTest.java:63]: It was really nice seeing the footprint no longer increases with entry count.

File: hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java [hunk 2]
```diff
@@ -81,32 +80,29 @@
...
-        leafKeys = new OAHashSet[leaves];
-        for (int i = 0; i < leaves; i++) {
-            leafKeys[i] = new OAHashSet<Object>(1);
-        }
-
-        initializeFootprint();
+        this.footprint = INT_SIZE_IN_BYTES * tree.length
+                // reference to the tree
+                + REFERENCE_COST_IN_BYTES
+                // reference to leafKeys array
```
Comment (blazember) [inline hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java:96]: I think we don't have this anymore since `leafKeys` is removed

Comment (mmedenjak) [reply hazelcast/src/main/java/com/hazelcast/wan/impl/merkletree/ArrayMerkleTree.java]: Removed.

**Comment by mmedenjak** [comment]
run-lab-run

**Comment by mmedenjak** [comment]
Thank you for the reviews, guys!

**Pull request merged**